Source/DafnyCore/Resolver/HigherOrderHeapAllocationCheckerConstructor.cs

```diff
@@ -0,0 +1,106 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+namespace Microsoft.Dafny;
+
+// This checker prevents the definition of non-terminating functions
+// by storing functions in memory (aka Landin's knots) during the 
+// creation of an object. This could also be prevented by doing 
+// a cardinality check on the type of class (similar to what is done
+// for algebraic datatypes) but at a fundamental level, the issue is
+// the creation of a Landin's knot, and it is easier and safer to catch
+// this way because of type parameters and traits.
+// Thanks to frame information, we need not reject all assignments of
+// functions to memory, but only the ones that are know to have a 
+// read frame.
+// To understand this checker, it is recommended to first look at
+// HigherOrderHeapAllocationChecker.
+class HigherOrderHeapAllocationCheckerConstructor : ASTVisitor<IASTVisitorContext> {
+  private readonly ErrorReporter reporter;
+
+  public HigherOrderHeapAllocationCheckerConstructor(ErrorReporter reporter) {
+    this.reporter = reporter;
+  }
+
+  public override IASTVisitorContext GetContext(IASTVisitorContext astVisitorContext, bool inFunctionPostcondition) {
+    return astVisitorContext;
+  }
+
+  // Occurs is a pure function that visits a type (rhs) to test
+  // for the presence of an heap-allocated type (Obj) left of
+  // an arrow. 
+  // This check could be relaxed but we keep it simple until
+  // we encounter a good use case for the more general check.
+  // Recall that a cycle can only be created using a function,
+  // which is why the test is different than a traditional
+  // cardinality test.
+  private bool Occurs(Type Obj, Type rhs, bool left) {
+    Type type = rhs.NormalizeExpandKeepConstraints();
+    if (type is BasicType) {
+      return false;
+    } else if (type is MapType) {
+      var t = (MapType)type;
+      return Occurs(Obj, t.Domain, left) || Occurs(Obj, t.Range, left);
+    } else if (type is SetType) {
+      var t = (SetType)type;
+      return Occurs(Obj, t.Arg, left);
+    } else if (type is CollectionType) {
+      var t = (CollectionType)type;
+      return Occurs(Obj, t.Arg, left);
+    } else {
+      var t = (UserDefinedType)type;
+      if (left && Obj.Equals(t)) {
```

Comment:
I _think_ this `.Equals` will resolve to the virtual method `object.Equals`. I don't remember if we override it in `Type` and then call the static method `Type.Equals`. If so, that would be pretty good. But I think you actually want `Type.Equal_Improved(Obj, t)`.

My memory is a bit blurry, but I _think_ that `Type.Equal_Improved` was introduced as a bug fix for `Type.Equals(_, _)`, and the intention had been to replace the latter with the improved one, but that the replacement never happened. Or, maybe these two methods do different things.